### PR TITLE
Replace `sys.exit` calls in `conda_build/metadata.py`

### PR DESCRIPTION
**Pull request by beeankha**
### Description

Replacing `sys.exit` call in a few functions in `conda_build/metadata.py` with the new `CondaBuildUserError` exception for better error handling, along with corresponding unit tests. 

The changes in this PR are separated out from work previously done by @kenodegard in #5255
Xref #4209 

### Checklist - did you ...

- [ ] ~~Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?~~
     
- [x] Add / update necessary tests?
- [ ] ~~Add / update outdated documentation?~~

### CODE REVIEW

**Review by travishathaway — APPROVED** [review]

File: conda_build/metadata.py [hunk 1]
```diff
@@ -46,18 +55,12 @@
...
     OutputDict = dict[str, Any]
     OutputTuple = tuple[OutputDict, "MetaData"]
 
-try:
-    import yaml
-except ImportError:
-    sys.exit(
-        "Error: could not import yaml (required to read meta.yaml "
-        "files of conda recipes)"
-    )
```
Comment (jezdez) [inline conda_build/metadata.py:60]: Was that basically an implicit dependency? Do we require it now in meta.yml?

File: conda_build/metadata.py [hunk 2]
```diff
@@ -350,16 +353,9 @@ def yamlize(data):
     try:
         return yaml.load(data, Loader=StringifyNumbersLoader)
     except yaml.error.YAMLError as e:
-        if "{{" in data:
-            try:
-                import jinja2
-
-                jinja2  # Avoid pyflakes failure: 'jinja2' imported but unused
-            except ImportError:
-                raise exceptions.UnableToParseMissingJinja2(original=e)
```
Comment (jezdez) [inline conda_build/metadata.py:359]: Just to understand, why are we removing this here? This looks like something 3rd party users might be relying on to identify why this is failing to parse

File: conda_build/metadata.py [hunk 3]
```diff
@@ -767,28 +761,18 @@ def _git_clean(source_meta):
...
 
     # make a copy of the input so we have no side-effects
     ret_meta = source_meta.copy()
-    # loop over the old versions
-    for key, has in zip(git_rev_tags[1:], has_rev_tags[1:]):
-        # update if needed
-        if has:
-            ret_meta[git_rev_tags[0]] = ret_meta[key]
-        # and remove
-        ret_meta.pop(key, None)
```
Comment (jezdez) [inline conda_build/metadata.py:791]: That seems like an unrelated change to this PR?

File: conda_build/metadata.py [hunk 4]
```diff
@@ -2024,17 +1997,6 @@ def _get_contents(
...
-            import jinja2
-        except ImportError:
-            print("There was an error importing jinja2.", file=sys.stderr)
-            print(
-                "Please run `conda install jinja2` to enable jinja template support",
-                file=sys.stderr,
-            )  # noqa
-            with open(self.meta_path) as fd:
-                return fd.read()
-
```
Comment (jezdez) [inline conda_build/metadata.py:2037]: Is jinja2 a dependency now?

File: conda_build/metadata.py (hunk 1 above)
Comment (dholth) [reply conda_build/metadata.py:60]: I'm surprised that it doesn't say "ruamel.yaml" but that package is certainly a regular dependency now.

Comment (beeankha) [reply conda_build/metadata.py:60]: It was previously an optional dependency but is currently [listed as a requirement in `meta.yml`](https://github.com/conda/conda-build/commit/f60aa9a3bee04e0d6d39166dba231786c3b14939#diff-0286c1d624d881984aef773a82433d955692c2fcbc773e7a6230fdb9602a23a3R31).

This particular code change is more "cleanup" vs "remove `sys.exit`" (per work originally done in [this PR](https://github.com/conda/conda-build/pull/5255/files#diff-a3f5613bda9366e31149c65731327047d4e29204c9e0508fd55416115b39a6bdL50-L57)), so I will revert this change; once the broken-out parts of the "remove `sys.exit` calls" work are merged, the [original PR](https://github.com/conda/conda-build/pull/5255) will get rebased and will mostly focus on cleanup.

File: conda_build/metadata.py (hunk 2 above)
Comment (beeankha) [reply conda_build/metadata.py:359]: Similar to `yaml`/`pyyaml`, `jinja2` used to be an optional dependency but is now [listed as a requirement in `meta.yml`](https://github.com/conda/conda-build/commit/c538eeaf4c23d0c3959a44f881997865f32d7707). However, since this is more "cleanup" vs "remove `sys.exit`", I will revert this change and leave it for the [original PR](https://github.com/conda/conda-build/pull/5255).

File: conda_build/metadata.py (hunk 3 above)
Comment (beeankha) [reply conda_build/metadata.py:791]: Yep, I'm going to revert this change and leave it for the [original "remove `sys.exit` calls" PR](https://github.com/conda/conda-build/pull/5255) (which got very unwieldy and is thus getting broken up into smaller chunks like this PR.

**Review by travishathaway — APPROVED** [review]
